crates/geoblock/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+        Self {
+            blocked_countries,
+            geoip,
+        }
+    }
+}
+
+impl<S, T> Layer<S> for GeoBlockLayer<T>
+where
+    T: GeoIpResolver,
+{
+    type Service = GeoBlock<S, T>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        GeoBlock::new(inner, self.geoip.clone(), self.blocked_countries.clone())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct GeoBlock<S, R>
+where
+    R: GeoIpResolver,
+{
+    inner: S,
+    blocked_countries: Vec<String>,
+    geoip: R,
+}
+
+impl<S, R> GeoBlock<S, R>
+where
+    R: GeoIpResolver,
+{
+    fn new(inner: S, geoip: R, blocked_countries: Vec<String>) -> Self {
+        Self {
+            inner,
+            blocked_countries,
+            geoip,
+        }
+    }
+
+    fn extract_ip<ReqBody>(&self, req: &Request<ReqBody>) -> Result<IpAddr, GeoBlockError> {
+        req.extensions()
+            .get::<ConnectInfo<SocketAddr>>()
+            .map(|ConnectInfo(addr)| addr.ip())
+            .ok_or(GeoBlockError::UnableToExtractIPAddress)
+    }
+
+    fn check_caller(&self, caller: IpAddr) -> Result<(), GeoBlockError> {
+        let geo_data = self
+            .geoip
+            .lookup_geo_data(caller)
+            .map_err(|_| GeoBlockError::UnableToExtractGeoData)?;
+
+        // TODO: let configure how to handle missing country
+        let country = geo_data
+            .country
+            .ok_or(GeoBlockError::CountryNotFound)?
+            .to_lowercase();
+
+        let is_blocked = self
+            .blocked_countries
+            .iter()
+            .any(|blocked_country| blocked_country == &country);
+
+        if is_blocked {
+            Err(GeoBlockError::BlockedCountry { country })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl<R, S, ReqBody, ResBody> Service<Request<ReqBody>> for GeoBlock<S, R>
+where
+    R: GeoIpResolver,
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Body + Default,
+{
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, ResBody>;
+    type Response = Response<ResBody>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        match self.extract_ip(&req) {
+            Ok(ip_addr) => match self.check_caller(ip_addr) {
+                Ok(_) => ResponseFuture::future(self.inner.call(req)),
+                Err(_e) => {
+                    let mut res = Response::new(ResBody::default());
+                    *res.status_mut() = StatusCode::UNAUTHORIZED;
+                    ResponseFuture::invalid_ip(res)
+                }
+            },
+            Err(_e) => {
+                let mut res = Response::new(ResBody::default());
+                *res.status_mut() = StatusCode::UNAUTHORIZED;
+                ResponseFuture::invalid_ip(res)
+            }
+        }
+    }
```

Comment:
Do we even need the `Error` enum then? It seems like it's only being used internally. I suggest ditching the `Error` and operating on `StatusCode`s directly:

```rust
fn check_caller(&self, req: &Request<Body>) -> Result<(), StatusCode> {
   // use  self.extract_ip() here
}
```

```rust
fn call(&mut self, req: Request<ReqBody>) -> Self::Future { 
    self.check_caller(&req)
        .map(|_| ResponseFuture::future(self.inner.call(req))
        .map_err(||code| ResponseFuture::error(code))
}

Reply:
I guess we may want to have it just to as a lower-level building block, e.g. if we implement a `warp` filter outside of this crate.

Reply:
The error is not public anymore, only used internally.